cranelift-faerie/src/backend.rs

```diff
@@ -179,6 +508,46 @@ impl Backend for FaerieBackend {
         // because `define` will take ownership of code, this is our last chance
         let code_length = code.len() as u32;
 
+        if let Some(ref mut frame_sink) = self.frame_sink {
+            if let Some(layout) = ctx.func.frame_layout.as_ref() {
+                let reg_mapper = DwarfRegMapper::for_isa(&self.isa);
+
+                let (cie, mut encoder) = frame_sink.cie_for(&layout.initial, &reg_mapper);
+
+                let mut fd_entry =
+                    FrameDescriptionEntry::new(frame_sink.address_for(name), code_length);
+
+                let mut frame_changes = vec![];
+                for ebb in ctx.func.layout.ebbs() {
+                    for (offset, inst, size) in
+                        ctx.func.inst_offsets(ebb, &self.isa.encoding_info())
+                    {
+                        if let Some(changes) = layout.instructions.get(&inst) {
+                            for change in changes.iter() {
+                                frame_changes.push((offset + size, change.clone()));
+                            }
+                        }
+                    }
+                }
+
+                frame_changes.sort_by(|a, b| a.0.cmp(&b.0));
+
+                let fde_insts = frame_changes
+                    .into_iter()
+                    .flat_map(|(addr, change)| encoder.translate(&change).map(|inst| (addr, inst)));
+
+                for (addr, inst) in fde_insts.into_iter() {
+                    fd_entry.add_instruction(addr, inst);
+                }
+
+                frame_sink.add_fde(cie, fd_entry);
+            } else {
+                // we have a frame sink to write .eh_frames into, but are not collecting debug
```

Comment:
Please panic here to make it easier to debug in that case.

Reply:
I realized this morning that there's a case where we will reach this arm but not be a bug: if the user of cranelift has a per-function `nounwind`-like attribute, a `nounwind` function wouldn't need unwind information collected, even though there still would be a frame sink for other functions. Right now it happens to be that Lucet is all-or-nothing on unwind info, but as a rustc backend for example I could see some function being opt-out.

Given that, would you rather a panic to identify the case for later discussion, or should I discard the `else` entirely?

Reply:
> but as a rustc backend for example I could see some function being opt-out.

I believe unwinding enabled/disabled is done per crate (using `-Cpanic=abort`), not per function.

Reply:
rustc has the unstable `#[unwind(allowed/aborts)]` to control per function, as well I think it has different requirements for `extern "C"`, but I don't know the details of that and it is under ongoing discussion.